File: go/candlestickpatterns/patterns/risingfallingthreemethods.go
package patterns

import (
	"zpano/candlestickpatterns/core"
	"zpano/fuzzy"
)

// RisingFallingThreeMethods detects a Rising/Falling Three Methods pattern:
// a five-candle continuation pattern.
//
// Uses TA-Lib logic: opposite-color check via color multiplication,
// real-body overlap (not full candle containment), sequential closes,
// 5th opens beyond 4th close.
//
// Category B: direction from 1st candle color (crisp sign).
func RisingFallingThreeMethods(cp *core.CandlestickPatterns) float64 {
	if !cp.Enough(5, cp.LongBody, cp.ShortBody) {
		return 0.0
	}

	b1 := cp.Bar(5)
	b2 := cp.Bar(4)
	b3 := cp.Bar(3)
	b4 := cp.Bar(2)
	b5 := cp.Bar(1)

	// Fuzzy: 1st long, 2nd-4th short, 5th long.
	muLong1 := cp.MuGreater(core.RealBodyLen(b1.O, b1.C), cp.LongBody, 5)
	muShort2 := cp.MuLess(core.RealBodyLen(b2.O, b2.C), cp.ShortBody, 4)
	muShort3 := cp.MuLess(core.RealBodyLen(b3.O, b3.C), cp.ShortBody, 3)
	muShort4 := cp.MuLess(core.RealBodyLen(b4.O, b4.C), cp.ShortBody, 2)
	muLong5 := cp.MuGreater(core.RealBodyLen(b5.O, b5.C), cp.LongBody, 1)

	// Determine color of 1st candle: +1 white, -1 black — crisp sign.
	color1 := 1.0
	if !core.IsWhite(b1.O, b1.C) {
		color1 = -1.0
	}

	// Color check: white, 3 black, white OR black, 3 white, black — crisp.
	c2 := 1.0
	if !core.IsWhite(b2.O, b2.C) {
		c2 = -1.0
	}
	c3 := 1.0
	if !core.IsWhite(b3.O, b3.C) {
		c3 = -1.0
	}
	c4 := 1.0
	if !core.IsWhite(b4.O, b4.C) {
		c4 = -1.0
	}
	c5 := 1.0
	if !core.IsWhite(b5.O, b5.C) {
		c5 = -1.0
	}

	if !(c2 == -color1 && c3 == c2 && c4 == c3 && c5 == -c4) {
		return 0.0
	}

	// 2nd to 4th hold within 1st: a part of the real body overlaps 1st range — crisp.
	if !(min(b2.O, b2.C) < b1.H && max(b2.O, b2.C) > b1.L &&
		min(b3.O, b3.C) < b1.H && max(b3.O, b3.C) > b1.L &&
		min(b4.O, b4.C) < b1.H && max(b4.O, b4.C) > b1.L) {
		return 0.0
	}

	// 2nd to 4th are falling (rising) — using color multiply trick — crisp.
	if !(b3.C*color1 < b2.C*color1 && b4.C*color1 < b3.C*color1) {
		return 0.0
	}

	// 5th opens above (below) the prior close — crisp.
	if !(b5.O*color1 > b4.C*color1) {
		return 0.0
	}

	// 5th closes above (below) the 1st close — crisp.
	if !(b5.C*color1 > b1.C*color1) {
		return 0.0
	}

	conf := fuzzy.TProductAll(muLong1, muShort2, muShort3, muShort4, muLong5)
	return color1 * conf * 100.0
}

File: ts/candlestick-patterns/patterns/rising-falling-three-methods.ts
/** Rising/Falling Three Methods pattern (5-candle continuation). */
import { CandlestickPatternsEngine } from '../core/engine.ts';
import { tProductAll } from '../../fuzzy/index.ts';
import { realBody, isWhite } from '../core/primitives.ts';

/**
 * Rising/Falling Three Methods: a five-candle continuation pattern.
 *
 * Uses TA-Lib logic: opposite-color check via color multiplication,
 * real-body overlap (not full candle containment), sequential closes,
 * 5th opens beyond 4th close.
 *
 * Category B: direction from 1st candle color (crisp sign).
 *
 * Returns:
 *     Continuous float in [-100, +100].
 */
export function risingFallingThreeMethods(cp: CandlestickPatternsEngine): number {
    if (!cp.enough(5, cp.longBody, cp.shortBody)) return 0.0;

    const b1 = cp.bar(5);
    const b2 = cp.bar(4);
    const b3 = cp.bar(3);
    const b4 = cp.bar(2);
    const b5 = cp.bar(1);

    // Fuzzy: 1st long, 2nd-4th short, 5th long.
    const muLong1 = cp.muGreaterCS(realBody(b1.o, b1.c), cp.longBody, 5);
    const muShort2 = cp.muLessCS(realBody(b2.o, b2.c), cp.shortBody, 4);
    const muShort3 = cp.muLessCS(realBody(b3.o, b3.c), cp.shortBody, 3);
    const muShort4 = cp.muLessCS(realBody(b4.o, b4.c), cp.shortBody, 2);
    const muLong5 = cp.muGreaterCS(realBody(b5.o, b5.c), cp.longBody, 1);

    // Determine color of 1st candle: +1 white, -1 black — crisp sign.
    const color1 = isWhite(b1.o, b1.c) ? 1.0 : -1.0;

    // Color check: white, 3 black, white  OR  black, 3 white, black — crisp.
    const c2 = isWhite(b2.o, b2.c) ? 1.0 : -1.0;
    const c3 = isWhite(b3.o, b3.c) ? 1.0 : -1.0;
    const c4 = isWhite(b4.o, b4.c) ? 1.0 : -1.0;
    const c5 = isWhite(b5.o, b5.c) ? 1.0 : -1.0;

    if (!(c2 === -color1 && c3 === c2 && c4 === c3 && c5 === -c4)) return 0.0;

    // 2nd to 4th hold within 1st: a part of the real body overlaps 1st range — crisp.
    if (!(Math.min(b2.o, b2.c) < b1.h && Math.max(b2.o, b2.c) > b1.l &&
        Math.min(b3.o, b3.c) < b1.h && Math.max(b3.o, b3.c) > b1.l &&
        Math.min(b4.o, b4.c) < b1.h && Math.max(b4.o, b4.c) > b1.l)) return 0.0;

    // 2nd to 4th are falling (rising) — using color multiply trick — crisp.
    if (!(b3.c * color1 < b2.c * color1 && b4.c * color1 < b3.c * color1)) return 0.0;
    // 5th opens above (below) the prior close — crisp.
    if (!(b5.o * color1 > b4.c * color1)) return 0.0;
    // 5th closes above (below) the 1st close — crisp.
    if (!(b5.c * color1 > b1.c * color1)) return 0.0;

    return color1 * tProductAll(muLong1, muShort2, muShort3, muShort4, muLong5) * 100.0;
}
